File: Graph/Java/Find Minimum Time to Reach Last Room I.java
// Problem link: https://leetcode.com/problems/find-minimum-time-to-reach-last-room-i/description/
// Time: O(n * m * log(n * m))
// Space: O(n * m)
class Solution {
    public int minTimeToReach(int[][] moveTime) {
        int rows = moveTime.length, cols = moveTime[0].length, rw, cl, row, col, dist, store[];
        int[][] distance = new int[rows][cols];
        PriorityQueue<int[]> PQ = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));
        int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

        for(int[] dis : distance) {
            Arrays.fill(dis, Integer.MAX_VALUE);
        }

        distance[0][0] = 0;
        PQ.offer(new int[] {0, 0, 0});

        while(!PQ.isEmpty()) {
            store = PQ.poll();
            dist = store[0];
            row = store[1];
            col = store[2];

            if(distance[row][col] < dist) {
                continue;
            }
            else if(row == rows - 1 && col == cols - 1) {
                return dist;
            }

            for(int[] dir : directions) {
                rw = dir[0] + row;
                cl = dir[1] + col;

                if(rw >= 0 && rw < rows && cl >= 0 && cl < cols && distance[rw][cl] > Math.max(dist, moveTime[rw][cl]) + 1) {
                    distance[rw][cl] = Math.max(dist, moveTime[rw][cl]) + 1;
                    PQ.offer(new int[] {distance[rw][cl], rw, cl});
                }
            }
        }

        return distance[rows - 1][cols - 1];
    }
}

File: Graph/Python/Find Minimum Time to Reach Last Room I.py
# Problem link: https://leetcode.com/problems/find-minimum-time-to-reach-last-room-i/description/
# Time: O(n * m * log(n * m))
# Space: O(n * m)
class Solution:
    def minTimeToReach(self, moveTime: List[List[int]]) -> int:
        rows, cols, PQ = len(moveTime), len(moveTime[0]), []
        directions = [[0, 1], [1, 0], [-1, 0], [0, -1]]
        distance = [[float('inf')] * cols for _ in range(rows)]

        distance[0][0] = 0
        heappush(PQ, (0, 0, 0))

        while PQ:
            dist, row, col = heappop(PQ)

            if dist > distance[row][col]:
                continue
            elif row == rows - 1 and col == cols - 1:
                return dist
            
            for dr, dc in directions:
                rw, cl = row + dr, col + dc

                if 0 <= rw < rows and 0 <= cl < cols and distance[rw][cl] > max(moveTime[rw][cl], dist) + 1:
                    distance[rw][cl] = max(moveTime[rw][cl], dist) + 1
                    heappush(PQ, (distance[rw][cl], rw, cl))
        
        return distance[rows - 1][cols - 1]
